### PR TITLE
ChartJs factory and Browserify support

### PR DESCRIPTION
**Pull request by jantimon**
This pull request adds the following features/fixes: 

 + [Turn the ChartJs library into a service](https://github.com/jtblin/angular-chart.js/compare/master...jantimon:master#diff-08b0d6c5641587816a2d854cc26def08R25)  
to support angular dependency injection
 + [Create a directive factory](https://github.com/jtblin/angular-chart.js/compare/master...jantimon:master#diff-08b0d6c5641587816a2d854cc26def08R6)  
  Your factory works awesome - however as it is privates it does not allow to add further directives (e.g. to support stacked bar charts or other custom / third party charts). This changes turns it into a public factory.
 + [Allow to set ChartJs configs via app.config](https://github.com/jtblin/angular-chart.js/compare/master...jantimon:master#diff-08b0d6c5641587816a2d854cc26def08R17)  
  to support the `angular.config` way to configure single charts or the entire library
 + [Add browserify support](https://github.com/jtblin/angular-chart.js/compare/master...jantimon:master#diff-08b0d6c5641587816a2d854cc26def08R25)

Hopefully all the red lines don't scare you. The pull request is actually way way smaller however because of the public factory feature I had to increase the indentation of many lines.
  
All unit/integration tests are still passing and I added an additional integration test.  
Please let me know what you think about this PR

### CODE REVIEW

**Comment by jtblin** [comment]
Thanks a ton @jantimon, it looks great. I see you preserved backward compatibility with global options `Chart.defaults.global` for colours which is neat (shall / can we log a deprecation message in the console when used to tell the users they should use the idiomatic angular config way now?). Though, I'm not sure the same is true for responsive and tooltip options, wouldn't that be overwritten here https://github.com/jtblin/angular-chart.js/compare/master...jantimon:master#diff-08b0d6c5641587816a2d854cc26def08R42? If that's the case can we check if it wasn't set to false? Given there's a bug with responsiveness when chart is in a hidden container, I am worry that could break people's charts.
I'll need to look a bit more into it over the weekend (as you said due to indentation, everything is red ;)) but after a quick review, that seems to be great additions.

**Comment by jtblin** [comment]
Just a quick note about

> Your factory works awesome - however as it is privates it does not allow to add further directives (e.g. to support stacked bar charts or other custom / third party charts). 

That's true but it was possible to use 3rd party chart by using the base type, there was even an example with the [stacked bar chart](https://github.com/jtblin/angular-chart.js/blob/master/examples/stacked-bars.html). That said, your way is much better :+1:

File: angular-chart.js [hunk 1]
```diff
@@ -1,243 +1,302 @@
...
+        ]
+      }
+    };
+
+    /**
+     * Allow to set global options during configuration
+     */
+    this.setOptions = function (type, options) {
+      // If no type was specified set option for the global object
+      if (options === undefined) {
```
Comment (jtblin) [inline angular-chart.js]: I prefer the `! options` notation.

File: test/fixtures/configure-line-chart.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+(function () {
+  'use strict';
+
+  var app = angular.module('line', ['chart.js']);
+
+  app.config(function (ChartJsProvider) {
+    // Configure all charts
+    ChartJsProvider.setOptions({
```
Comment (jtblin) [inline test/fixtures/configure-line-chart.js]: Neat, could you add an example to the readme as well please?

File: dist/angular-chart.js [hunk 3]
```diff
@@ -1,243 +1,302 @@
...
+   *   ChartJsProvider.setOptions({ responsive: true });
+   *   ChartJsProvider.setOptions('Line', { responsive: false });
+   * })))
+   */
+  function ChartJsProvider() {
+    var Chart = (typeof window !== 'undefined' && window.Chart) ||
+      (typeof require !== 'undefined' && require('chart.js'));
+
+    if (typeof Chart !== 'function') {
+      throw 'Failed to load chart.js';
```
Comment (jtblin) [inline dist/angular-chart.js]: Need to throw an error i.e. `new Error('Failed to load chart.js')`.

File: angular-chart.js [hunk 4]
```diff
@@ -1,243 +1,302 @@
...
+   *   ChartJsProvider.setOptions({ responsive: true });
+   *   ChartJsProvider.setOptions('Line', { responsive: false });
+   * })))
+   */
+  function ChartJsProvider() {
+    var Chart = (typeof window !== 'undefined' && window.Chart) ||
+      (typeof require !== 'undefined' && require('chart.js'));
+
+    if (typeof Chart !== 'function') {
+      throw 'Failed to load chart.js';
```
Comment (jtblin) [inline angular-chart.js]: Need to throw an error i.e. `new Error('Failed to load chart.js')`.

File: angular-chart.js [hunk 5]
```diff
@@ -1,243 +1,302 @@
...
-
-        if (typeof window.G_vmlCanvasManager === 'object' && window.G_vmlCanvasManager !== null) {
-          if (typeof window.G_vmlCanvasManager.initElement === 'function') {
-            window.G_vmlCanvasManager.initElement(elem[0]);
-          }
-        }
+    .provider('ChartJs', ChartJsProvider)
+    .factory('ChartJsFactory', ['ChartJs', ChartJsFactory])
+    .directive('chartBase',
+      ['ChartJsFactory', function (ChartJsFactory) { return new ChartJsFactory(); }])
```
Comment (jtblin) [inline angular-chart.js]: Feel like we could dry the code here by having a function returning the array with the type as parameter. Although a factory factory starts feeling very java-ish ;)

File: angular-chart.js [hunk 6]
```diff
@@ -1,243 +1,302 @@
...
-          if (! chartType) return;
+    this.$get = function () {
+      return ChartJs;
+    };
+  }
 
-          if (chart) {
-            if (canUpdateChart(newVal, oldVal)) return updateChart(chart, newVal, scope);
-            chart.destroy();
+  function ChartJsFactory(ChartJs) {
```
Comment (jtblin) [inline angular-chart.js]: Minor nitpick, add a space before round braces in function declarations (I need to add the corresponding `jscs` rule).

**Comment by jantimon** [comment]
Wow - thank's for all your feedback.
Hopefully I will find some time tomorrow to solve the remaining issues.

File: angular-chart.js (hunk 5 above)
Comment (jantimon) [inline angular-chart.js]: well that's only an annotation for the minification and can be automated by https://www.npmjs.com/package/gulp-ng-annotate

File: test/fixtures/configure-line-chart.js (hunk 2 above)
Comment (jantimon) [inline test/fixtures/configure-line-chart.js]: Done.

**Comment by jantimon** [comment]
@jtblin I worked through most of your points.
One thing that's still open is the backwards compatibility.
You are right  `Chart.defaults.global.responsive = false` is broken - but I have a solution in my mind.
I'll try to fix this tonight. :)

File: angular-chart.js (hunk 4 above)
Comment (jantimon) [inline angular-chart.js]: Done.

File: angular-chart.js (hunk 6 above)
Comment (jantimon) [inline angular-chart.js]: Done. (and added the jscs rule)

**Comment by jantimon** [comment]
@jtblin I worked through the entire feedback and fixed all mentioned points.
In order to keep the full `Chart.defaults.global` compatibility I had to set the default options like they were set in 0.6.0.
Would love to get further feedback.

**Comment by jantimon** [comment]
Uses now a standard amd factory.
I guess this pull request is ready for merge. 
What do you think?

**Comment by jtblin** [comment]
Awesome, thanks a lot! Yes ready to merge, I'll do that tonight.

**Comment by jantimon** [comment]
Awesome :+1:

**Comment by jgodi** [comment]
Do you mind cutting a release with this? I was just looking over this PR and would very much love to start using this :)

**Comment by jtblin** [comment]
It's merged so you can take a fork from master now, I'll be releasing
shortly :)

**Comment by jantimon** [comment]
Just a short side node - I forgot to bump the bower version.

**Pull request merged**